### ISSUE TITLE
Data pipeline failed without retry

### ISSUE THREAD

**Issue by zuston** [Kimahriman/hdfs-native#154]
![image](https://github.com/user-attachments/assets/f8fe52a7-820d-4cba-9d5a-77afbab0e2e5)

When using the hdfs-native crate, I encountered the data pipeline failed. If this happened by the problem datanode, do we need to request a new block location? 

Could you help look this problem? @Kimahriman

**Comment by zuston** [comment]
After digging the vannila hdfs code, it looks it will replace bad datanode if pipeline failed.

![image](https://github.com/user-attachments/assets/6462115f-bbcb-4314-8b77-2da6650d67ae)

**Comment by Kimahriman** [comment]
Yeah the write path is the least resilient part of the library right now. I need to take a deeper look to understand exactly how the Java writer handles lost data nodes, and then figure out how I could write a test for something like that.

If you're not already, I'd currently recommend retrying the whole write from scratch if you still have access to all the data you are trying to write.

**Comment by Kimahriman** [comment]
Based on a quick look, it seems like the logic is generally:
- If a DataNode fails, either the one it is talking to or any of the others being replicated to, simply remove that DataNode from the block being written and continue on
- If certain conditions are true, add a new DataNode to the block. Specifically the default conditions are in this comment:
```
DEFAULT condition:
   *   Let r be the replication number.
   *   Let n be the number of existing datanodes.
   *   Add a new datanode only if r >= 3 and either
   *   (1) floor(r/2) >= n or (2) the block is hflushed/appended.
```
So for example with replication of 3, a new DataNode would only get added if two DataNodes fail during writing that block.

The first one is simpler and I could look into trying to add that, though still not exactly sure how I would test it. That should make things more resilient than the current behavior which is simply failing on the first DataNode error.

The second is a bit more complex and requires copying the already written data to a new replica before continuing, and probably wouldn't be something I could tackle anytime soon.

**Comment by zuston** [comment]
Thanks for your quick reply, I found another error status when appending file. Please see this: 

![IMG_2676](https://github.com/user-attachments/assets/22a67e5f-2c70-4fb9-ba9d-52a4c5d41088)
 
And I think the first one solution is good for my append case.

**Comment by zuston** [comment]
> Yeah the write path is the least resilient part of the library right now. 

Yes, it looks unstable especially in a busy hdfs cluster. 

If you have any improvement patch, i’m willing to test this. Almostly, i could reproduce it by my cases.

**Comment by Kimahriman** [comment]
> Thanks for your quick reply, I found another error status when appending file. Please see this:

This would be when a data node the block is replicating to fails. The connection drop would be when the one you are talking to dies, so I think they're effectively the same issue

**Comment by zuston** [comment]
After simply looking this article https://blog.cloudera.com/understanding-hdfs-recovery-processes-part-2/, I think the bad data node replacement will not trigger the original data resend. Once client found the bad data node, it will build the new pipeline through the namenode, and then resend the data from last acked packet. @Kimahriman

**Comment by zuston** [comment]
Thanks for your code reference. Let me take a deep look.

**Comment by zuston** [comment]
> The data is re-replicated by the client on node replacement.

From digging the DataStreamer.java, the replicated transfer is from the healthy datanode -> replacement datanodes. The client only trigger the recovery pipeline creating to back fill the data to replacement node.

BTW, if ignore the bad nodes and do nothing, the missing blocks will exist. So If we want to have a resilient writing process, the datanode replacement may need to be supported

**Comment by zuston** [comment]
Attach the hadoop code reference: https://github.com/apache/hadoop/blob/96572764921706b1fecaf064490457d36d73ea6e/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataNode.java#L3593

**Comment by Kimahriman** [comment]
> From digging the DataStreamer.java, the replicated transfer is from the healthy datanode -> replacement datanodes. The client only trigger the recovery pipeline creating to back fill the data from healthy node to replacement node.

Yeah looks like it just makes an RPC call to trigger the replication, just a little bit of extra complexity on there. I think that would still be follow on work to just getting the write to continue with fewer DataNodes on failures.

> BTW, if ignore the bad nodes and do nothing, the missing blocks will exist. So If we want to have a resilient writing process, the datanode replacement may need to be supported

My understanding is even if a single node (or multiple but not all) in a pipeline are lost, if you successfully write the block to at least one DataNode, HDFS will re-replicate it after the fact as it will see it is under replicated. I think the main reason for replicating as part of the pipeline recovery is just to make that a little more resilient. i.e. if in the end all but one DataNodes fail in your pipeline, and you "successfully" finish your write, but then that one DataNode immediately dies, your write was successful but now you are missing that block.

**Comment by zuston** [comment]
And do you have time or passion to improve this? I think I just start learn from this part and not sure I have enough knowledge to develop this feature.

**Comment by zuston** [comment]
![image](https://github.com/user-attachments/assets/90c64a91-e98a-4c9e-9610-dd004fa77726)

another possible error message

**Comment by zuston** [comment]
![image](https://github.com/user-attachments/assets/fac7982a-17e5-4c83-b0d6-108622d3be7a)

**Comment by zuston** [comment]
> Hmm that's very odd, might be unrelated to writing data. What version of HDFS are you running? Also if you can replicate those having some debug logs would be helpful to see what's going on.

HDFS with 3.2.2 version. 

> Also if you can replicate those having some debug logs would be helpful to see what's going on.

It's OK, I will attach it if I having more detailed logs from the namenode or datanodes.

**Comment by Kimahriman** [comment]
I think I have an initial version of at least handling the failures correctly if you want to test it out in your environment: https://github.com/Kimahriman/hdfs-native/pull/164

**Comment by zuston** [comment]
> I think I have an initial version of at least handling the failures if you want to test it out in your environment: #164

Yes, I will test this in the later 3 days. Thanks for fixing this.
